File: frontend/customer-app/src/types/book.ts
export interface Book {
  id: string;
  isbn: string;
  title: string;
  description: string;
  price: number;
  stock_quantity: number;
  publisher_id: string;
  publication_date: string;
  language: string;
  pages: number;
  format: string;
  cover_image_url?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  authors?: Author[];
  categories?: Category[];
  publisher?: Publisher;
}

export interface Author {
  id: string;
  name: string;
  bio?: string;
  birth_date?: string;
  created_at: string;
  updated_at: string;
}

export interface Category {
  id: string;
  name: string;
  slug: string;
  parent_id?: string;
  created_at: string;
  updated_at: string;
}

export interface Publisher {
  id: string;
  name: string;
  country?: string;
  website?: string;
  created_at: string;
  updated_at: string;
}

export interface BookFilters {
  title?: string;
  author?: string;
  author_id?: string;
  category_id?: string;
  min_price?: number;
  max_price?: number;
  page?: number;
  page_size?: number;
  limit?: number;
  offset?: number;
}

export interface BooksResponse {
  books: Book[];
  total: number;
  page: number;
  page_size: number;
}

File: services/catalog-service/internal/domain/book.go
package domain

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Book represents a book in the catalog
type Book struct {
	ID              uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	ISBN            string         `gorm:"type:varchar(13);uniqueIndex;not null" json:"isbn"`
	Title           string         `gorm:"type:varchar(500);not null" json:"title"`
	Description     string         `gorm:"type:text" json:"description"`
	Price           float64        `gorm:"type:decimal(10,2);not null" json:"price"`
	StockQuantity   int            `gorm:"type:integer;default:0" json:"stock_quantity"`
	PublisherID     *uuid.UUID     `gorm:"type:uuid" json:"publisher_id,omitempty"`
	Publisher       *Publisher     `gorm:"foreignKey:PublisherID" json:"publisher,omitempty"`
	CoverImageURL   string         `gorm:"type:text" json:"cover_image_url"`
	PublicationDate *time.Time     `gorm:"type:date" json:"publication_date,omitempty"`
	Language        string         `gorm:"type:varchar(50);default:'English'" json:"language"`
	PageCount       int            `gorm:"type:integer" json:"page_count"`
	Authors         []Author       `gorm:"many2many:book_authors;" json:"authors,omitempty"`
	Categories      []Category     `gorm:"many2many:book_categories;" json:"categories,omitempty"`
	CreatedAt       time.Time      `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt       time.Time      `gorm:"autoUpdateTime" json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// TableName overrides the table name for GORM
func (Book) TableName() string {
	return "books"
}
